hepstats/hypotests/toyutils.py

```diff
@@ -0,0 +1,356 @@
+        """
+        return self._nlls
+
+    @property
+    def ntoys(self):
+        """
+        Returns the number of toys.
+        """
+        return len(self.bestfit)
+
+    def add_entries(self, bestfit, nll_bestfit, nlls):
+        """
+        Add new result entries.
+
+        Args:
+            * bestfit (`numpy.array`): best fitted values of the POI
+            * nll_bestfit (`numpy.array`): NLL  evaluated at the best fitted values of the POI
+            * nlls (Dict(`POI`, `numpy.array`)): NLL  evaluated at the best fitted values of the POI
+        """
+        if not all(k in nlls.keys() for k in self.poieval):
+            missing_keys = [k for k in self.poieval if k not in nlls.keys()]
+            raise ValueError(f"NLLs values for {missing_keys} are missing.")
+
+        nentries = bestfit.size
+        assert nll_bestfit.size == nentries
+        assert all(a.size == nentries for a in nlls.values())
+
+        self._bestfit = np.concatenate([self.bestfit, bestfit])
+        self._nll_bestfit = np.concatenate([self.nll_bestfit, nll_bestfit])
+
+        self._nlls = {p: np.concatenate([v, nlls[p]]) for p, v in self.nlls.items()}
+
+    def to_dict(self):
+        """
+        Returns dictionnary of the toy results.
+        """
+        ret = {"poi": self.poigen.name, "bestfit": self.bestfit}
+        ret["nlls"] = {n.value: nll for n, nll in self.nlls.items()}
+        ret["genvalue"] = self.poigen.value
+        ret["evalvalues"] = self.poieval.values
+        ret["nlls"]["bestfit"] = self.nll_bestfit
+        return ret
+
+
+class ToysManager(ToysObject):
+
+    def __init__(self, input, minimizer, sampler=base_sampler, sample=base_sample):
+        """Class handling the toy generation and fit, results are stored in `ToyResult` instances stored
+            themselves in a dictionnary.
+
+            Args:
+                input : loss or fit result
+                minimizer : minimizer to use to find the minimum of the loss function
+                ntoysnull (int, default=100): minimum number of toys to generate for the null hypothesis
+                ntoysalt (int, default=100): minimum number of toys to generate for the alternative hypothesis
+                sampler : function used to create sampler with models, number of events and
+                    floating parameters in the sample. Default is `hepstats.fitutils.sampling.base_sampler`.
+                sample : function used to get samples from the sampler.
+                    Default is `hepstats.fitutils.sampling.base_sample`.
+        """
+
+        super(ToysManager, self).__init__(input=input, minimizer=minimizer, sampler=sampler, sample=sample)
+        self._toys = {}
+
+    def get_toyresult(self, poigen, poieval):
+        """
+        Getter function.
+
+        Args:
+            poigen (POI): POI used to generate the toys
+            poieval (POIarray, optional): POI values to evaluate the loss function
+
+        Returns:
+            `Toys`
```

Comment:
Should the "Returns" not be a list of `ToyResult` instead? (Apologies if I'm being slow, here.)